src/kernel.jl

```diff
@@ -84,14 +84,179 @@ function set_arg!(k::Kernel, idx::Integer, arg::LocalMem)
+bytes. For example, a float4 variable will be aligned to a 16-byte boundary, a char2 variable will
+be aligned to a 2-byte boundary.
+For 3-component vector data types, the size of the data type is 4 * sizeof(component). This
+means that a 3-component vector data type will be aligned to a 4 * sizeof(component)
+boundary. The vload3 and vstore3 built-in functions can be used to read and write, respectively,
+3-component vector data types from an array of packed scalar data type.
+A built-in data type that is not a power of two bytes in size must be aligned to the next larger
+power of two. This rule applies to built-in types only, not structs or unions.
+The OpenCL compiler is responsible for aligning data items to the appropriate alignment as
+required by the data type. For arguments to a `__kernel` function declared to be a pointer to a
+data type, the OpenCL compiler can assume that the pointee is always appropriately aligned as
+required by the data type. The behavior of an unaligned load or store is undefined, except for the
+vloadn, vload_halfn, vstoren, and vstore_halfn functions defined in section 6.12.7. The vector
+load functions can read a vector from an address aligned to the element type of the vector. The
+vector store functions can write a vector to an address aligned to the element type of the vector.
+"""
+function cl_alignement(x)
+    is_cl_inbuild(x) ? inbuild_alignement(x) : cl_sizeof(x)
+end
+function inbuild_alignement{T}(::Type{T})
+    x = sizeof(T)
+    ispow2(x) ? x : nextpow2(x)
+end
+inbuild_alignement{T}(x::T) = inbuild_alignement(T)
+
+
+function advance_aligned(offset, alignment)
+    (offset == 0 || alignment == 0) && return 0
+    if offset % alignment != 0
+        npad = ((div(offset, alignment) + 1) * alignment) - offset
+        offset += npad
+    end
+    offset
+end
+
+
+"""
+Sizeof that considers OpenCL alignement. See cl_alignement
+"""
+Base.@pure function _cl_sizeof{T}(::Type{T}, offset = 0)
+    align, size = if is_cl_inbuild(T) || nfields(T) == 0
+        align, size = inbuild_alignement(T), sizeof(T)
+        if offset == 0
+            return align
+        end
+        align, size
+    else
+        nextoffset = offset
+        for field in fieldnames(T)
+            xT = fieldtype(T, field)
+            nextoffset = _cl_sizeof(xT, nextoffset)
+        end
+        size = nextoffset - offset
+        size, size
+    end
+    offset = advance_aligned(offset, align)
+    offset += size
+    offset
+end
+
+Base.@generated function cl_sizeof{T}(x::T)
+    size = _cl_sizeof(T, 0)
+    :($size)
+end
+Base.@generated function cl_sizeof{T}(x::Type{T})
+    size = _cl_sizeof(T, 0)
+    :($size)
+end
+
+"""
+Converts a Julia type to a correctly padded OpenCL type.
+Turns into a no-op when no convert needed
+"""
+@generated function aligned_convert(x)
+    offset = Ref(0); elements = []; fields = []
+    _aligned_convert!(x, offset, elements, fields, :x)
```

Comment:
Couldn't `_aligned_convert` simply return `offset, elements, fields` instead of having them passed and mutated?